src/dataflow/decode/mod.rs

```diff
@@ -217,47 +217,63 @@ where
     K: DecoderState + 'static,
     V: DecoderState + 'static,
 {
-    stream.unary(
-        Exchange::new(|x: &((Vec<u8>, _), _)| (x.0).0.hashed()),
-        &op_name,
-        move |_, _| {
-            move |input, output| {
-                input.for_each(|cap, data| {
-                    let mut session = output.session(&cap);
-                    for ((key, payload), aux_num) in data.iter() {
-                        if key.is_empty() {
-                            error!("{}", "Encountered empty key");
-                            continue;
+    stream
+        .aggregate::<_, (Vec<u8>, Option<i64>), _, _, _>(
+            |_key, val, agg| {
+                if let Some(new_offset) = val.1 {
+                    if let Some(offset) = agg.1 {
+                        if offset < new_offset {
+                            *agg = val;
                         }
-                        match key_decoder_state.decode_key(key) {
-                            Ok(key) => {
-                                if payload.is_empty() {
-                                    session.give((key, None, *cap.time()));
-                                } else {
-                                    value_decoder_state.give_key_value(
-                                        key,
-                                        payload,
-                                        *aux_num,
-                                        &mut session,
-                                        *cap.time(),
-                                    );
-                                }
+                    } else {
+                        *agg = val;
+                    }
+                }
+            },
+            |key, agg| (key, agg),
+            |key| key.hashed(),
+        )
```

Comment:
This certainly does a thing, but I think we need to be clear about what it does. That should probably be in the form of some comments, that explain either (or both) what this does, and what `decode_key_values_inner` does, because I this changes its semantics. It's probably a "good change" for upserts, but if this method is only meant to be used for upserts we should give it an appropriate name.

Reply:
For example, this first operator seems to drop records that do not have the optional `i64`, which isn't what was done before. It would be great to explain the intent of the operator, as I'm not currently sure it does what it needs to do (as likely will be future people who might need to tweak or understand it).

If it shouldn't be discarding, you should just be able to keep and compare by the offset, for which `None` is the least element.

Reply:
The other potentially weird thing that it does is collapse values down by *capability* rather than logical timestamp. This might be ok for common uses of upserts, but it means that should we receive inputs that look like
```
(key, val1, time1)
(key, val2, time2)
(key, val3, time2)
```
we may lose the ability to recover the fine distinction of going from `val1` to `val3`. I understand at the moment that there is a 1:1 correspondence between timely capabilities and materialized logical timestamps, but differential doesn't require this and we may eventually want to remove it.